server/api/user.py

```diff
@@ -51,10 +56,64 @@ def onPostUser(event):
     attachUserPermissions(userResponse)
 
 
-def attachUserApi():
+def getUserEmail(user):
+    User = ModelImporter.model('user', 'isic_archive')
+    user = User.load(user['id'], force=True, exc=True)
+    return user['email']
+
+
+@access.user
+@describeRoute(
+    Description('Request permission to create datasets.'))
+def requestCreateDatasetPermission(params):
+    User = ModelImporter.model('user', 'isic_archive')
+    Group = ModelImporter.model('group')
+    currentUser = getCurrentUser()
+    resp = {}
+    if User.canCreateDataset(currentUser):
+        resp['message'] = 'Dataset Contributor access granted.',
+        resp['extra'] = 'hasPermission'
+    else:
+        # Request that user join group
+        groupName = 'Dataset Contributors'
+        group = Group.findOne({'name': groupName})
+        if not group:
+            raise RestException('Could not load group: %s' % groupName)
+        resp['message'] = 'Dataset Contributor access requested. Please wait ' \
+            'for an administrator to approve your request.'
+        requests = Group.getFullRequestList(group)
+        ids = [request['id'] for request in requests]
+        if not currentUser['_id'] in ids:
+            Group.joinGroup(group, currentUser)
+
+            # Send email to group moderators and administrators
+            groupAcl = Group.getFullAccessList(group)
+            groupModerators = [user for user
+                               in groupAcl['users']
+                               if user['level'] >= AccessType.WRITE]
+            emails = [getUserEmail(user) for user in groupModerators]
+            host = mail_utils.getEmailUrlPrefix()
+            html = mail_utils.renderTemplate(
+                'datasetContributorRequest.mako',
+                {
+                    'user': currentUser,
+                    'group': group,
+                    'host': host,
+                })
+            mail_utils.sendEmail(
+                to=emails,
+                subject='ISIC Archive Dataset Contributor Request',
+                text=html)
+
+    return resp
+
+
+def attachUserApi(user):
     events.bind('rest.get.user/authentication.after',
                 'onGetUserAuthentication', onGetUserAuthentication)
     events.bind('rest.get.user/me.after',
                 'onGetUserMe', onGetUserMe)
     events.bind('rest.post.user.after',
                 'onPostUser', onPostUser)
+    user.route('POST', ('requestCreateDatasetPermission',),
```

Comment:
This route name is a little wordy, but I can't think of a better alternative right now. I don't think this will be an important part of our documented API surface, so I'm fine with changing it later.